### PR TITLE
Upstream Merge [Q2/24]

### PR DESCRIPTION
**Pull request by stefankoncarevic**

### CODE REVIEW

File: external/llvm-project/mlir/lib/Conversion/TosaToLinalg/TosaToLinalgNamed.cpp [hunk 1]
```diff
@@ -1130,7 +1130,7 @@ void mlir::tosa::populateTosaToLinalgNamedConversionPatterns(
   }
   patterns->add<
       // clang-format off
-      ConvConverter<tosa::Conv2DOp, linalg::Conv2DNhwgcGfhwcOp, linalg::Conv2DNhwgcGfhwcQOp>,
+      ConvConverter<tosa::Conv2DOp, linalg::Conv2DNhwgcGfhwcOp, linalg::Conv2DNgchwGfchwQOp>,
```
Comment (krzysz00) [inline external/llvm-project/mlir/lib/Conversion/TosaToLinalg/TosaToLinalgNamed.cpp]: Wait hold on I think this is a bug?

Comment (stefankoncarevic) [reply external/llvm-project/mlir/lib/Conversion/TosaToLinalg/TosaToLinalgNamed.cpp]: You are right, I noticed it and fixed it in the next commit eaa86317f917052c6e9ab14742871e8189750182

**Comment by stefankoncarevic** [comment]
Problem with running premerge-checks.py script.
When I run **git-clang-format --diff origin/develop**, I get the following output:
/home/skoncare/mlir-dev/rocMLIR/external/llvm-project/clang/include/clang/Format/.clang-format:1:1: error: Unknown value for BasedOnStyle: clang-format
BasedOnStyle: clang-format
^
Error reading /home/skoncare/mlir-dev/rocMLIR/external/llvm-project/clang/include/clang/Format/.clang-format: Invalid argument
error: clang-format -lines=378:396 -lines=416:416 -lines=422:423 -lines=489:497 -lines=752:764 -lines=817:818 -lines=839:839 -lines=1170:1170 -lines=2634:2634 -lines=2656:2656 -lines=3033:3033 -lines=3089:3089 -lines=3618:3618 -lines=3653:3653 -lines=3656:3657 -lines=3661:3661 -lines=3695:3698 -lines=3711:3712 -lines=3714:3714 -lines=3718:3721 -lines=4049:4049 -lines=4141:4141 -lines=4153:4153 -lines=4189:4189 -lines=4319:4319 -lines=4671:4671 -lines=4807:4807 -lines=4811:4811 -lines=4876:4876 -lines=4932:4932 -lines=4964:4965 -lines=5248:5248 -lines=5265:5265 -lines=5390:5393 external/llvm-project/clang/include/clang/Format/Format.h failed

I also tried to add --ignore-external but it didn't help, I get the same error.
Upon further investigation, I noticed that the errors are related to four pull requests that made changes to the Format.h file. Here is the list of the relevant commits:
https://github.com/ROCm/llvm-project/commit/236b3e1aad45e2bab8ede0da6397b7b01f9cc9d8
https://github.com/ROCm/llvm-project/commit/1c58208d899285318c89e069268145c85ec33368
https://github.com/ROCm/llvm-project/commit/dba2aa265c5f2ef774c2812cf6ffdea8dd784e09
https://github.com/ROCm/llvm-project/commit/2fbc9f217e5fe8db8444a87dbd7138a768b8aa85

At this point, I do not know exactly what the problem is or how to resolve it. If anyone knows or thinks they can help me overcome this issue, it would be greatly appreciated.

Currently, I am building a new docker image where I update everything related to clang to version 18 to see if that helps resolve the issue. If this does not work, one option might be to revert these changes from the pull request for the upstream, but this is not a permanent solution.

**Comment by krzysz00** [comment]
Re clang-format, looks like the commit responsible is https://github.com/llvm/llvm-project/pull/69814 , which added the new style, and our copy of clang-format doesn't have that commit yet.

I think a safe temporary measure would be deleting the files that have a `BasedOnStyle: clang-format` in them for now, since we don't intend to build, let alone modify, the clang-format source code.

**Comment by krzysz00** [comment]
Now, review notes:
- Can you grab upstream's `llvm/lib/Target/AMDGPU/SIInsertHardClauses.cpp`? It's got the better fix for the hard clause issue
- Grab upstream's `llvm/lib/Target/ARM/ARMISelLowering.cpp ` for our sanity (we're not targetting ARM, but our version of the file has a merge conflict marker in it)
- @giuseros Is there an attempt to upstream the packed RTZ work ongoing?
- In `mlir/include/mlir/Conversion/Passes.td `, there are merge conflict markers (take upstream's) in `ConvertNVVMToLLVMPass`
  - Same file, in `ArithToAMDGPUConversionPass`, there are minor documentation wording issues (mainly around `description` and `option<"saturateFP8Truncf"`. Don't remove our new options, just use the wording that landed upstream
 - Stray comma after the definiton of `cluster_size` in GPUFuncOp in `mlir/include/mlir/Dialect/GPU/IR/GPUOps.td`
 - Drop `setInvariantLoadMetedataCode` from `mlir/Dialect/LLVMIR/LLVMOpBase.td` and `mlir/Dialect/LLVMIR/LLVMOps.td` - the mechanism for handling this changed during upstream review and that code is dead
 - Fix the position of the data layout string in `mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp` and check the `#include`s to see if they're any different from upstream's (aside from the math-to-llvm, let's keep that in how it is)

**Review by krzysz00 — APPROVED** [review]

**Comment by krzysz00** [comment]
@stefankoncarevic Please revert that change where you go randomly `clang-format`ing `external/llvm-project/` That's why we have `ignoreExternalLinting` in CI

**Review by krzysz00 — CHANGES_REQUESTED** [review]
I'm blocking this PR because you ran `clang-format` indiscriminately over upstream code, creating a whole lot of diff noise going forward. Please use `ignoreExternalLinting` instead

**Comment by stefankoncarevic** [comment]
@krzysz00 The commit before https://github.com/ROCm/rocMLIR/pull/1547/commits/2412b715f2453bce1235c8829507afd3e9698c33 reported an error on CI for the premerge script with --ignore-external flag. This is the link to CI:
http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-1547/17/pipeline/199/
The errors refer to the external folder, so I had to do a clang format for the script to execute successfully.
Is it possible that the problem is in the premerge script?
Maybe this is also related to the Docker file clang format update?

**Comment by krzysz00** [comment]
@stefankoncarevic  Yes, this means that the `ignoreExternalLinting` option isn't working correctly and you need to fix it

**Comment by krzysz00** [comment]
The point is, we shouldn't be carrying any patches to directories like `clang/` or `llvm/` unless they're actual temporary workarounds

**Comment by stefankoncarevic** [comment]
That was my thought as well. Was in a hurry so this seemed as an apt solution. Will fix the script. Thanks!

**Comment by stefankoncarevic** [comment]
The issue with the script only occurs on the CI. When I run the script locally with the --ignore-external flag, everything completes successfully. The problem likely stems from a recent change in the Docker file, where for some reason the clang-format returns an error. I've added a few print statements and initiated the CI to identify the problem, but due to the machines being busy, I'm still waiting for it to run.

**Comment by jerryyin** [comment]
> The issue with the script only occurs on the CI. When I run the script locally with the --ignore-external flag, everything completes successfully. The problem likely stems from a recent change in the Docker file, where for some reason the clang-format returns an error. I've added a few print statements and initiated the CI to identify the problem, but due to the machines being busy, I'm still waiting for it to run.

I'm not sure this is the right approach to use the CI to debug. You should be able to:
 - Create a new docker following recent changes
 - Replay locally and get the failure reproduced
 - Triage from there

**Comment by krzysz00** [comment]
Revert [38b9692](https://github.com/ROCm/rocMLIR/pull/1547/commits/38b969262fbc218c96490322c4ec2c7157e8f662) too

**Review by krzysz00 — APPROVED** [review]

**Pull request merged**